client/ayon_core/pipeline/workfile/workfile_template_builder.py

```diff
@@ -506,55 +506,61 @@ def build_template(
             keep_placeholders (bool): Add flag to placeholder data for
                 hosts to decide if they want to remove
                 placeholder after it is used.
-            create_first_version (bool): create first version of a workfile
-            workfile_creation_enabled (bool): If True, it might create
-                                              first version but ignore
-                                              process if version is created
+            create_first_version (bool): Create first version of a workfile.
+                 When set to True, this option initiates the saving of the
+                 workfile for an initial version. It will skip saving if
+                 a version already exists.
+            workfile_creation_enabled (bool): Whether we are creating a new
+                workfile. If False, we assume we explicitly want to build the
+                template in our current scene. Otherwise we only build if the
+                current file is not an existing saved workfile but a "new"
+                file.
 
         """
-        if any(
-            value is None
-            for value in [
-                template_path,
-                keep_placeholders,
-                create_first_version,
-            ]
+
+        # Get default values if not provided
+        if (
+                template_path is None
+                or keep_placeholders is None
+                or create_first_version is None
         ):
-            template_preset = self.get_template_preset()
-            if template_path is None:
-                template_path = template_preset["path"]
+            preset = self.get_template_preset()
+            template_path: str = template_path or preset["path"]
             if keep_placeholders is None:
-                keep_placeholders = template_preset["keep_placeholder"]
+                keep_placeholders: bool = preset["keep_placeholder"]
             if create_first_version is None:
-                create_first_version = template_preset["create_first_version"]
-
-        # check if first version is created
-        created_version_workfile = False
-        if create_first_version:
-            created_version_workfile = self.create_first_workfile_version()
+                create_first_version: bool = preset["create_first_version"]
 
-        # if first version is created, import template
-        # and populate placeholders
+        # Build the template if current workfile is a new unsaved file
+        # (that's detected by checking if it returns any current filepath)
         if (
-            create_first_version
-            and workfile_creation_enabled
-            and created_version_workfile
+                # If not a workfile creation, an explicit load template
+                # was requested, so we always want to build the template
```

Comment:
I still don't understand why that would be the case?

Reply:
Here's the execution flow: 
 
No. 1 Explicit Build: 
when triggering this button
![image](https://github.com/user-attachments/assets/899bd92c-08e4-4ba4-afd2-fdb83de62643)

It runs `build_workfile_template()` where `workfile_creation_enabled` argument will fall to `False`.
which tells the builder: hey, don't create any workfiles. just perform a regular build.

No. 2 `on_new()` callback
when starting a new scene, we runs `start_workfile_template_builder()` 
which runs the build function with the argument set to true,
`build_workfile_template(workfile_creation_enabled=True)`  which tells the builder: hey it's fine to create a workfile. after performing the regular build

Reply:
With those use cases in mind, an API interface like this one would make more sense to me:
```
   def build_template(
        self,
        template_path=None,
        level_limit=None,
        keep_placeholders=None,
        create_first_version=False,
        force_build=False,
    ):
```
where `force_build` is set to True only on the number 1 use case

Reply:
it would also help with the readability to move this part into its own variable:
```
is_new_scene = self.host.get_current_workfile() is None
```
so at the end it reads like:
```python
if is_new_scene or force_build:
```
much easier to read than negations